### DISCUSSION TITLE
ShredOS x86_64 build error- due to "failed to generate boot.vfat"

### DISCUSSION THREAD

**Discussion by ermibb** [PartialVolume/shredos.x86_64#127] [Q&A]
When I rebuild Shredos,on Neon OS I get the error like on screenshot:
![image](https://user-images.githubusercontent.com/62342832/220832221-ce37f705-c3d5-459b-8382-adad61751436.png)
So, how I can solve this type of problem?
Please help me!
Thank you!

**Comment by PartialVolume** [comment]
Have you made any changes from the vanilla ShredOS or changed any configurations or added additional programs?

**Reply by ermibb** [reply]
I have not entered any program in the original Shredos, I simply rebuild the program based on the following steps:
You can build shredos using the following commands. This example build was compiled on KDE Neon (Ubuntu 20.04).
Install the following prerequisite software first. Without this software, the make command will fail

$ sudo apt install git
$ sudo apt install build-essential   pkg-config   automake   libncurses5-dev   autotools-dev   libparted-dev   dmidecode   coreutils   smartmontools
$ sudo apt-get install libssl-dev
$ sudo apt-get install libelf-dev
$ sudo apt-get install mtools

Download the ShredOS source using the git clone command, build ShredOS and write to a USB memory device.

$ git clone https://github.com/PartialVolume/shredos.x86_64.git (or shredos.i686.git for 32bit)
$ cd shredos
$ mkdir package/shredos
$ touch package/shredos/Config.in
$ make clean
$ make shredos_defconfig
$ make
$ ls output/images/shredos*.img
$ cd output/images
$ dd if=shredos-20200412.img of=/dev/sdx (20200412 will be the day you compiled, sdx is the USB flash dri

When I run "make " command, almost after a hour and >10 GB shredos.x86_64 folder size, the command promot shows the above screenshot error.

**Comment by PartialVolume** [comment]
If you have added any additional programs or changed the configuration that results in an increase in ShredOS size, then you need to change a value in ../board/shredos/genimage.cfg. Change the value size from `size = 80000000` to `size = 90000000`. You will have to experiment a bit. If you still get the error increase to 100000000 etc, increasing by 10M.

Then just run the `make` command on its own. It shouldn't be necessary to rebuild from scratch as shown below.
```
# It shouldn't be necessary to do this ... just run make

$ git clone https://github.com/PartialVolume/shredos.x86_64.git
$ cd shredos
$ mkdir package/shredos
$ touch package/shredos/Config.in
$ make clean
$ make shredos_defconfig
$ make
```

 Here's the ../board/shredos/genimage.cfg file.
```
image boot.vfat {
        vfat {
                extraargs = "-F 32 -n SHREDOS"
                file boot/grub/grub.cfg     { image = 'grub.cfg' }
                file boot/shredos           { image = 'bzImage' }
                file EFI/BOOT/bootx64.efi   { image = 'bootx64.efi' }
                file EFI/BOOT/grub.cfg      { image = 'grub.cfg' }
                file shredos.ico           { image = 'shredos.ico' }
                file README.txt            { image = 'README.txt' }
                file autorun.inf           { image = 'autorun.inf' }
        }

        size = 80000000
}

image shredos.img {
        hdimage {
        }

        partition boot {
              in-partition-table = "no"
              image = "boot.img"
              offset = 0
              size = 512
        }

        partition grub {
              in-partition-table = "no"
              image = "grub.img"
              offset = 512
        }

        partition vfat {
                partition-type = 0xC
                image = "boot.vfat"
                offset = 646656
        }
}

**Reply by ermibb** [reply]
I made rebuilding process of  shredos using Neon OS in virtualbox, may be the error come from this?

Why I needed to rebuild  shredos?
Because I needed the program for work, I had the idea of changing the name(shredos  to other name), modify the instruction of how it works, and modifying it in my own language. So, Based on this, what do you advise me to do? 

Thanks 
Thank you very much for your cooperation
I look forward to your reply

**Reply by PartialVolume** [reply]
I would still increase the size value to 90000000. The existing value leaves only about 2MB in the fat partition when compiled on my system so the existing value doesn't leave much room for variations in the size when compiled on different systems.

Bearing in mind ShredOS pulls it's source code from the original sources, to change languages you would need to substantially patch nwipe and as nwipe is updated fairly often that sounds like an awful lot of work.

If you are serious about this I would first look at nwipes code base and work out how multi language support could be added. By doing it like this you wouldn't have to keep revising your patches for every new release.

How familiar are you with working with C code?

**Reply by PartialVolume** [reply]
getext sounds like a possibility of adding language support to nwipe.  https://levelup.gitconnected.com/how-to-make-a-multi-language-localized-application-in-c-8862e1d1f64f

and https://www.gnu.org/software/gettext/

**Reply by PartialVolume** [reply]
Using a more structured method like [gettext ](https://www.gnu.org/software/gettext/) would enable others to add their own languages to nwipe via .po files.

where shredos pulls its nwipe code from: [nwipe source](https://github.com/martijnvanbrummelen/nwipe)

**Comment by PartialVolume** [comment]
@ermibb There is a bug in mcopy v4.0.32 (fixed in v4.0.42) that causes the "Internal error: size too big" error. The fix is too use a v4.0.42 or edit the make file and remove the -b option from the mcopy command. The -b option is just a optimisation option so doesn't matter if it's removed.

**Comment by PartialVolume** [comment]
If your distro does not contain a later mtools package then the latest version can be obtained from [here](https://www.gnu.org/software/mtools/#downloads). I would suggest you uninstall the repository version before installing the new version.

See ShredOS readme [#issues-that-you-may-get-when-building-shredos](https://github.com/PartialVolume/shredos.x86_64#issues-that-you-may-get-when-building-shredos)
